### ISSUE TITLE
Critical error: incorrect passing of a pointer to an element of an array of structures

### ISSUE THREAD

**Issue by leon-ts** [X-Sharp/XSharpPublic#1132]
There is an array of structures. If you pass a pointer to a structure element to a function, then the compiler creates an incorrect address offset (without regard to the size of the structure).

**X#:** 2.13.2.2

**Example:**
```
FUNCTION Start() AS VOID STRICT

	LOCAL DIM st[3] IS TEST_STRUCT
	
	st[1].x := 1
	st[1].y := 1
	TestFunc(@st[1])
	
	st[2].x := 2
	st[2].y := 2
	TestFunc(@st[2])
	
	st[3].x := 3
	st[3].y := 3
	TestFunc(@st[3])
	
	WAIT

	RETURN

VOSTRUCT TEST_STRUCT
	MEMBER x AS INT
	MEMBER y AS INT
END VOSTRUCT

FUNCTION TestFunc(p AS TEST_STRUCT) AS VOID STRICT
	? p.x
	? p.y
	RETURN
```	

**ILSpy decompiled:**
![screenshot1](https://user-images.githubusercontent.com/64518775/191028235-e5bd5fb0-a659-4722-b5c9-831e17e0decf.png)

**Test solution:**
[TestConsoleApplication.zip](https://github.com/X-Sharp/XSharpPublic/files/9599249/TestConsoleApplication.zip)

P.S. Unfortunately, I was not able to detect this error in a timely manner. Is it possible to fix it with some urgent patch (before the end of this week)?

**Comment by RobertvanderHulst** [comment]
Leonid,
I'll try to fix this asap.
But I have no idea which change has broken this, so it could take a while to hunt this down.
What was the last build that you used, in which this worked ?

For now you could try this:
```
   LOCAL DIM st[3] IS TEST_STRUCT
    LOCAL p as TEST_STRUCT
    p := @st
	st[1].x := 1
	st[1].y := 1
	//TestFunc(@st[1])
	TestFunc(p)

	st[2].x := 2
	st[2].y := 2
	TestFunc(@st[2])
        p += 1
	TestFunc(p)


	st[3].x := 3
	st[3].y := 3
	//TestFunc(@st[3])
	p += 1
	TestFunc(p)

```

**Comment by leon-ts** [comment]
Robert,
The latest build of X# I worked with was 2.9.1.1.
Modification of the code on my part, unfortunately, is hardly possible. There are a lot of such places in the program, especially in my graphic libraries (arrays _winRECT, _winPOINT, etc.). I don't even know how to find them. There are hundreds of them, the names of the structures are different.

**Comment by leon-ts** [comment]
Robert,
In order not to distract you too much from other tasks, I will write here my thoughts about this problem.
I think the problem here is when casting to a pointer to byte:
`(byte*)ptr + 1`
In this case, the pointer increment is also incremented by one byte. If you remove this cast, then the pointer to the size of the structure will be incremented automatically.

From the Microsoft documentation (C#):

> For a pointer p of type T* and an expression n of a type implicitly convertible to int, uint, long, or ulong, addition and subtraction are defined as follows:
> - Both p + n and n + p expressions produce a pointer of type T* that results from adding n * sizeof(T) to the address given by p.
> - The p - n expression produces a pointer of type T* that results from subtracting n * sizeof(T) from the address given by p.

I made an example in C# to test the changes:
```
using XSharp.Internal;

namespace CSharpTestConsoleApplication
{
	class Program
	{
		static unsafe void Main(string[] args)
		{
			TEST_STRUCT[] st = new TEST_STRUCT[3];
			fixed (TEST_STRUCT* ptr = &st[0])
			{
				st[0].x = 1;
				st[0].y = 1;
				TestFunc(ptr);
				st[1].x = 2;
				st[1].y = 2;
				//TestFunc((TEST_STRUCT*)((byte*)ptr + 1));
				TestFunc(ptr + 1);
				st[2].x = 3;
				st[2].y = 3;
				//TestFunc((TEST_STRUCT*)((byte*)ptr + 2));
				TestFunc(ptr + 2);
				XSharp.RT.Functions._wait();
			}
		}
		public unsafe static void TestFunc(TEST_STRUCT* p)
		{
		}
	}
}

[VoStruct(8, 4)]
public struct TEST_STRUCT
{
	public int x;
	public int y;
}
```

As you can see, I removed the typecasts when incrementing the pointer:
```
TestFunc(ptr + 1);
TestFunc(ptr + 2);
```
And it worked, the pointer began to be transferred correctly.

**Comment by leon-ts** [comment]
Robert,
If it helps in your search, here's what the code looked like compiled into X# 2.9.1.1:

![screenshot2](https://user-images.githubusercontent.com/64518775/191246848-aa3bda16-49eb-4bea-b599-beeefe2899dc.png)

**Comment by RobertvanderHulst** [comment]
Leon,
Thanks for the info.
I will be able to look at this tomorrow (Wednesday 21) at the earliest.

**Comment by leon-ts** [comment]
Robert,
Thanks!

**Comment by RobertvanderHulst** [comment]
Leonid,
We made a change in this area because there was a report about the DIM local not properly being pinned.
So we changed the code to use the generated Pinned TEST_STRUCT* instead of the address of the local.
Assuming this pointer is named ptr this should have generated
`TestFunc(ptr + 1)`
and 
`TestFunc(ptr + 2)`
Unfortunately we had typed the (generated) binary operator as an `BinaryOperatorKind.IntAddition` and not a `BinaryOperatorKind.PointerAndIntAddition`. This resulted in the cast to byte.
I have changed this now and it seems to work.
I'll run some internal tests now and when the tests run fine, then I will send you a set of updated compiler assemblies.
The generated IL now looks like this (in C# mode)
```
                TEST_STRUCT[] st = new TEST_STRUCT[3];
		fixed (TEST_STRUCT* ptr = &st[0])
		{
			st[0].x = 1;
			st[0].y = 1;
			TestFunc(ptr);
			st[1].x = 2;
			st[1].y = 2;
			TestFunc(ptr + 1);
			st[2].x = 3;
			st[2].y = 3;
			TestFunc(ptr + 2);
		}
```

The source code for the last call in assembly looks like this

```
                        IL_007f: ldloc.1
			IL_0080: ldc.i4.2
			IL_0081: conv.i
			IL_0082: sizeof TEST_STRUCT
			IL_0088: mul
			IL_0089: add
			IL_008a: call void Tester.Exe.Functions::TestFunc(valuetype TEST_STRUCT*)
```
So it takes the pointer and adds 2 x the size of TEST_STRUCT, exactly as you would expect.
I think this is also much cleaner than the original code
